### ISSUE TITLE
gfx_defines! dynamic/data driven alternative?

### ISSUE THREAD

**Issue by greenmughal** [gfx-rs/gfx#1325]
Is it possible that vertex,  pipeline and other stuff that macro gfx_defines! builds can also be created using data files or on the fly? 

Regards

**Comment by msiglreith** [comment]
Hi!
it's possible but a bit cumbersome. There are basically 2 options for doing this

1. In the `render` layer by defining the Init and MetaData struct yourself and implement the necessary traits (e.g https://github.com/ebkalderon/amethyst/blob/renderer/amethyst_renderer/src/pipe/effect/pso.rs). This allows you to re-use our `Encoder` and `Bundle` for drawing.

2. Going into the `core` layer and manually fill the [Descriptor](https://docs.rs/gfx/0.16.0/gfx/pso/struct.Descriptor.html). Then you need to use raw [CommandBuffer](https://docs.rs/gfx/0.16.0/gfx/trait.CommandBuffer.html) interface, which requires you to also bind the resources. We currently don't have a backend-agnostic way for creating raw command buffers iirc, so if you only use a specific backend you just need to lookup the corresponding function in the Factory impl or write an trait extension if you want to do it backend-agnostic. (See https://github.com/gfx-rs/gfx/blob/master/src/lib.rs, which does it for encoder creation. The trait impls also show how the command buffers are created)

**Comment by kvark** [comment]
The question boils down to implementing a "raw" PSO type that can be constructed dynamically. IIRC, @ebkalderon was looking at this - did you find anything?

**Comment by ebkalderon** [comment]
@kvark I did exactly what @msiglreith mentioned above. I first attempted the second approach, where I would manually fill out a `Descriptor` and use that to create a `RawPipelineState` struct; unfortunately, that barred me from using the convenient `Encoder` struct and resulted in a lot of boilerplate which the typed `PipelineState` would have handled for me anyway.

As such, I settled on the former approach since it was easier to implement and also let me reuse existing code that expects a typed `PipelineState`.

**Comment by kvark** [comment]
@ebkalderon right. Do you think your raw PSO can be re-used by other projects?

**Comment by ebkalderon** [comment]
I don't see why not! It doesn't really depend on anything specific to Amethyst. Perhaps the only thing that's missing is a builder for more convenient initialization (see the `Effect` builder that I started working on).

**Comment by greenmughal** [comment]
@ebkalderon Can you  point me to sample code of first approach? Thanks

**Comment by ebkalderon** [comment]
@greenmughal See the source file linked by @msiglreith in his post above.
